File: day01/main.rb
def solve_part_one(file)
  return parse_file_and_determine_calories_per_elf(file)
    .max(1)
    .sum
end

def solve_part_two(file)
  return parse_file_and_determine_calories_per_elf(file)
    .max(3)
    .sum
end

def parse_file_and_determine_calories_per_elf(file)
  elves = []
  calories = 0

  File.readlines(file).each do |line|
    if line.chomp.empty?
      elves.append(calories)
      calories = 0
    else
      calories += line.to_i
    end
  end

  return elves
end

if __FILE__ == $0
  puts "Part 1: #{solve_part_one("day01/input.txt")}" # 72070
  puts "Part 2: #{solve_part_two("day01/input.txt")}" # 211805
end

File: day01/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)

func solvePartOne(path string) int {
	elves := parseInput(path)
	return elves[0]
}

func solvePartTwo(path string) int {
	elves := parseInput(path)
	return elves[0] + elves[1] + elves[2]
}

func parseInput(path string) []int {
	elves := make([]int, 0)
	calories := 0

	file, _ := os.Open(path)
	fscanner := bufio.NewScanner(file)
	for fscanner.Scan() {
		line := fscanner.Text()
		if len(line) == 0 {
			elves = append(elves, calories)
			calories = 0
		} else {
			snack, _ := strconv.Atoi(line)
			calories += snack
		}
	}
	sort.Sort(sort.Reverse(sort.IntSlice(elves)))
	return elves
}

func main() {
	fmt.Println("Part 1:", solvePartOne("day01/input.txt"))
	fmt.Println("Part 2:", solvePartTwo("day01/input.txt"))
}
